pystiche/papers/ulyanov_et_al_2016/core.py

```diff
@@ -0,0 +1,207 @@
+        optimizer = get_optimizer(transformer)
+
+        lr_scheduler = ulyanov_et_al_2016_lr_scheduler(
+            optimizer, impl_params=impl_params,
+        )
+
+    if num_epochs is None:
+        if impl_params:
+            if instance_norm:
+                num_epochs = 25
+            else:
+                num_epochs = 10 if stylization else 5
+        else:
+            num_epochs = 10
+
+    style_transform = ulyanov_et_al_2016_style_transform(
+        impl_params=impl_params, instance_norm=instance_norm
+    )
+    style_transform = style_transform.to(device)
+    preprocessor = ulyanov_et_al_2016_preprocessor()
+    preprocessor = preprocessor.to(device)
+    style_image = style_transform(style_image)
+    style_image = preprocessor(style_image)
+    style_image = batch_up_image(style_image, loader=content_image_loader)
+    criterion.set_style_image(style_image)
+
+    if stylization:
+
+        def criterion_update_fn(input_image, criterion):
+            criterion.set_content_image(preprocessor(input_image))
+
+    else:
+
+        def criterion_update_fn(input_image, criterion):
+            pass
+
+    return default_transformer_epoch_optim_loop(
+        content_image_loader,
+        transformer,
+        criterion,
+        criterion_update_fn,
+        num_epochs,
+        device=device,
+        lr_scheduler=lr_scheduler,
+        quiet=quiet,
+        logger=logger,
+        log_fn=log_fn,
+    )
+
+
+def ulyanov_et_al_2016_stylization(
+    input_image: torch.Tensor,
+    transformer: Union[nn.Module, str],
+    impl_params: bool = True,
+    instance_norm: bool = False,
+    stylization: bool = True,
+):
+    device = input_image.device
+    if isinstance(transformer, str):
+        style = transformer
+        transformer = ulyanov_et_al_2016_transformer(
+            style=style,
+            impl_params=impl_params,
+            instance_norm=instance_norm,
+            stylization=stylization,
+        )
+        if instance_norm or not impl_params:
+            transformer = transformer.eval()
+        transformer = transformer.to(device)
+
+    with torch.no_grad():
+        content_transform = ulyanov_et_al_2016_content_transform(
+            impl_params=impl_params, instance_norm=instance_norm
+        )
+        content_transform = content_transform.to(device)
+        input_image = content_transform(input_image)
+        postprocessor = ulyanov_et_al_2016_postprocessor()
+        postprocessor = postprocessor.to(device)
+        output_image = transformer(input_image)
+        output_image = postprocessor(output_image)
+
+    return output_image.detach()
+
+
+def ulyanov_et_al_2016_texture_generation(
+    input: Union[Tuple[int, int], torch.Tensor],
+    transformer: Union[nn.Module, str],
+    impl_params: bool = True,
+    instance_norm: bool = False,
+    stylization: bool = True,
+):
+    if isinstance(input, torch.Tensor):
+        device = input.device()
```

Comment:
The attribute `Tensor.device()` is not callable. If you simply want to extract the `device` use `input.device`.